### ISSUE TITLE
Datatable results count is incorrect when checkAccess is set

### ISSUE THREAD

**Issue by eric-burel** [VulcanJS/Vulcan#2066]
Hi,

Currently the DataTable loadMore button is displayed when `count < totalCount` where count is the number of item retrieved, and totalCount the number of item in the database.

This method is incorrect when a `checkAccess` method is defined. The correct comparison would be `count + filteredOut < totalCount`, where `filteredOut` is the number of results eliminated when you call `checkAccess` (or `beforeFilterCount < total` where `beforeFilterCount` is the number of results before filtering).

For example, if you have 10 items but user can see only 8, all data are loaded when 8 results are printed and 2 results are filtered out. But the comparison is `8 < 10`, so the user thinks there are 2 more results to load while we already know that he actually can't access them.

You can get this count easily in the `default_resolver` and return it alongside the total. However I don't manage to add new fields into the `withMulti` hoc results in addition to `data.totalCount`.

**Comment by SachaG** [comment]
If I understand correctly, I think this is fundamentally unfixable because it would require calling `checkAccess` on every document in the database… I think the best we can do is add an option to disable the document count altogether when you need to.

**Comment by eric-burel** [comment]
You can't compute it in the db, however you have access of the number of filtered document when you call `checkAccess`. 
So basically you can't compute the total of accessible documents, but you can still remember the total of document your refused to the user, and send it to him. 
Hopefully that's enough for a loadMore button, because you know you are done the total of document you have + the total you filtered is equal to the total of document in the db. 
You don't need to filter all the document at once, you only filter the documents retrieved by the user, and you can still have a correct behaviour.

For pagination, it is true that you can't take the number of accessible documents into account when you compute the number of pages. So you may require 10 document per page but only get 4 (or zero) after filtering, and indeed there is not solution to this because you can't filter documents.

To sum it up the use of a custom JS function for document level permission management will never be fully compatible with pagination, but it could be compatible with a `loadMore`pattern.

**Comment by meteorplus** [comment]
@eric-burel can you show me an example file where you use the checkAccess ?

thanks

**Comment by eric-burel** [comment]
Hi @meteorplus,

[See this link](https://github.com/lbke/vulcan-validation-workflows/blob/master/lib/modules/setupDocumentValidation.js#L60-L78). This is an example function that make unvalidated data unavailable, except if you actually created those data. It works on any collection.

To answer you on `totalCount`, `totalCount` is the number of item your MongoDB `find` query would have returned if you did not set a limit. So, if you don't filter your data, `totalCount` = number of documents in your database. If you filter your data, `totalCount` = number of document in your database that match a filter criteria.

This `totalCount` is computed at the **database** level. However, `checkAccess` is applied later, at the **server** level. But currently, the number of data filtered by `checkAccess` is not saved, so there is absolutely no way to tell how many documents it discarded.

There are alternative solutions, so you should actually barely need this value.  [I've updated the doc about this](https://github.com/VulcanJS/vulcan-docs/blob/master/source/groups-permissions.md) (see Backend vs Frontend). But still, there are a few use case where it could be necessary when `checkAccess` is very complex and you can't simplify it.

**Comment by saravanabalagi** [comment]
@eric-burel More than I would need the `totalCount` filtered by `checkAccess`, I want others NOT have access to this value. I have collections readable only by the ones who created it, but now exposing the total entries is what I would consider a data leak. Any chance we can block access to the `totalCount` altogther?

**Comment by eric-burel** [comment]
Hi @saravanabalagi, your request makes sense indeed, can you open a separate issue for this? I think a solution would be to add a `disableTotal` option when creating the default resolver that blocks totalCount resolution for the collection. (https://github.com/VulcanJS/Vulcan/blob/devel/packages/vulcan-core/lib/modules/default_resolvers.js)
Currently you can simply write your own custom resolver to solve this: if you don't provide a resolver for the totalCount it's as if it was not available.
